New autocoder employing XML definition file format

cdjackson opened the pull request:
This PR is a rewrite of the autocoder. Primarily, it changes from an ```MD``` format to an ```XML``` format. The MD file was getting unmanageable in size as more clusters are added, and it is also more difficult to parse and add new constructs.

This may change the name of some methods due to the way the file is parsed - eg ```Iascie``` is now formatted in methods/fields as ```IasCie``` and ```ZoneID``` is now ```ZoneId``` so this may break some implementations (sorry!).

There may also be some changes to commands and attributes as these are now marked with a direction (eg wether an attribute resides in the client or server). This is one of the drivers behind this change - ie to allow a future enhancement which supports different attributes for clients and servers, as required by some clusters.

Currently this PR adds attributes for the server. This needs to be resolved as the OTA is implemented as a client on the device, so this will cause it to not work right now!

@hsudbrock FYI. I still need to test this - it generates code that is pretty much the same as before, but there might be some changes...

Signed-off-by: Chris Jackson <chris@cd-jackson.com>

cdjackson commented:
This pull request **introduces 2 alerts** and **fixes 1** when merging caa6133f6a2d52edc98f4042b6570f5e555e8645 into da475facf3c569a07db132d6d781609498b1b1ae - [view on LGTM.com](https://lgtm.com/projects/g/zsmartsystems/com.zsmartsystems.zigbee/rev/pr-b1b84b5b282cd73a183f13d8efafa383bd25b675)

**new alerts:**

* 1 for Useless comparison test
* 1 for Dereferenced variable may be null

**fixed alerts:**

* 1 for Useless comparison test

---

*Comment posted by [LGTM.com](https://lgtm.com)*

cdjackson commented:
This pull request **introduces 2 alerts** and **fixes 10** when merging 06ad4e217b5a292bf318ba348c892f32fd6b7654 into 2cb131576419c25164147f0c3fd3df340a695a3e - [view on LGTM.com](https://lgtm.com/projects/g/zsmartsystems/com.zsmartsystems.zigbee/rev/pr-c33c3c7c0aab261b1a7dbbd4736f550478fc76fe)

**new alerts:**

* 1 for Useless comparison test
* 1 for Dereferenced variable may be null

**fixed alerts:**

* 7 for Useless comparison test
* 1 for Unused format argument
* 1 for Missing format argument
* 1 for Useless null check

---

*Comment posted by [LGTM.com](https://lgtm.com)*

cdjackson commented:
This pull request **introduces 2 alerts** and **fixes 10** when merging ccd804a6dadaf2b6247dca2c5f46e481e5a1c913 into 2cb131576419c25164147f0c3fd3df340a695a3e - [view on LGTM.com](https://lgtm.com/projects/g/zsmartsystems/com.zsmartsystems.zigbee/rev/pr-7f5cf56aaf1ad3129bbc98cc0d4de3dca40a28f2)

**new alerts:**

* 1 for Useless comparison test
* 1 for Dereferenced variable may be null

**fixed alerts:**

* 7 for Useless comparison test
* 1 for Unused format argument
* 1 for Missing format argument
* 1 for Useless null check

---

*Comment posted by [LGTM.com](https://lgtm.com)*

t-8ch commented:
Is it intentional, that the generated code is commited to the repo?
I would have expected the code to be generated at build-time during the `generate-sources` phase.

t-8ch commented:
FYI, there is a nice library to support with java code generation: https://github.com/square/javapoet
The current generation seems fine, just to make you aware of it.

t-8ch commented on com.zsmartsystems.zigbee.autocode/src/main/java/com/zsmartsystems/zigbee/autocode/FileCopy.java:
```diff
@@ -0,0 +1,108 @@
+/**
```
This class seems not to be used. Is something missing from the PR?

cdjackson commented:
> Is it intentional, that the generated code is commited to the repo?

I have been thinking about changing to this concept, but it's always been this way, and I've not bothered to change it. If you want to provide a PR, and it works ok, then that would be great :)

> FYI, there is a nice library to support with java code generation: https://github.com/square/javapoet

Thanks. The autocoders in this PR aren't new so I don't really anticipate changing this.

> This class seems not to be used. Is something missing from the PR?

Thanks - looks like I moved the functions out of the class at some point in the past, so it's not used. I've removed it.

cdjackson commented:
This pull request **introduces 2 alerts** and **fixes 10** when merging e6df9bdc289e3c5e951e054e44e00571c0730d1d into 2cb131576419c25164147f0c3fd3df340a695a3e - [view on LGTM.com](https://lgtm.com/projects/g/zsmartsystems/com.zsmartsystems.zigbee/rev/pr-65cc04218d0a10cae1ad5ba7aca14188d4fb8666)

**new alerts:**

* 1 for Useless comparison test
* 1 for Dereferenced variable may be null

**fixed alerts:**

* 7 for Useless comparison test
* 1 for Unused format argument
* 1 for Missing format argument
* 1 for Useless null check

---

*Comment posted by [LGTM.com](https://lgtm.com)*

cdjackson commented:
This pull request **introduces 2 alerts** and **fixes 10** when merging 4be095b95b7ceca7c27b3d1002b4b8b14b463fce into 263da3135543c4f0bf1061fd15be551ca541ac9a - [view on LGTM.com](https://lgtm.com/projects/g/zsmartsystems/com.zsmartsystems.zigbee/rev/pr-3648efdcf67097df5aa9953dcf3683123af4a10a)

**new alerts:**

* 1 for Useless comparison test
* 1 for Dereferenced variable may be null

**fixed alerts:**

* 7 for Useless comparison test
* 1 for Unused format argument
* 1 for Missing format argument
* 1 for Useless null check

---

*Comment posted by [LGTM.com](https://lgtm.com)*

hsudbrock commented:
@cdjackson I have started testing with the code from this PR; so far, things look good.

We discussed that it could be good to include these changes together with the changes necessary for the manufacturer-specific clusters, so I will simply rebase my PR for the manufacturer-specific clusters on this PR.

cdjackson commented:
Thanks Henning - sounds good. 

Sent from my iPhone

> On 22 Feb 2019, at 08:34, Henning Sudbrock <notifications@github.com> wrote:
> 
> @cdjackson I have started testing with the code from this PR; so far, things look good.
> 
> We discussed that it could be good to include these changes together with the changes necessary for the manufacturer-specific clusters, so I will simply rebase my PR for the manufacturer-specific clusters on this PR.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

hsudbrock commented:
Chris, I am currently developing in openHAB based on this branch, also for testing it some more. What would be helpful is support for client-side attributes, as otherwise discoveries always run into a NPE (the node property discoverer tries to read the FW version, which fails as this is a client-side attribute that is not yet written into the attribute list of the `ZclOtaUpgradeCluster`).

So, if you happen to push updates to this branch, this would be helpful for me :) (But I can work around this in the time being...)

hsudbrock commented on com.zsmartsystems.zigbee.autocode/src/main/java/com/zsmartsystems/zigbee/autocode/ZigBeeZclCommandGenerator.java:
```diff
@@ -0,0 +1,239 @@
+                out.println("    public boolean isTransactionMatch(ZigBeeCommand request, ZigBeeCommand response) {");
+                if (command.response.matchers.isEmpty()) {
+                    out.println("        return (response instanceof " + command.response.command + ")");
+                    out.println("                & ((ZdoRequest) request).getDestinationAddress().equals((("
```
Small typo: This should be `&&` here (one `&` is missing).

cdjackson commented:
This pull request **introduces 2 alerts** and **fixes 8** when merging 5ada35fc64342f2cae584faa685e1a0d4acff530 into ec8c99e3a715b374b1cf332f1865bb7ad6ed87d5 - [view on LGTM.com](https://lgtm.com/projects/g/zsmartsystems/com.zsmartsystems.zigbee/rev/pr-242681f70904c82c37a646e78d25d233cfcaa5d1)

**new alerts:**

* 1 for Useless comparison test
* 1 for Dereferenced variable may be null

**fixed alerts:**

* 7 for Useless comparison test
* 1 for Useless null check

---

*Comment posted by [LGTM.com](https://lgtm.com)*

cdjackson commented:
This pull request **introduces 2 alerts** and **fixes 8** when merging 2de321fc1f3f1f9093e306dc69d9e28e8d6deb3c into 921b979c74ec5be71e515fdc42f1cbe2ecd85d0e - [view on LGTM.com](https://lgtm.com/projects/g/zsmartsystems/com.zsmartsystems.zigbee/rev/pr-bedf876d24c317971334ee7de88f41f1b7c11896)

**new alerts:**

* 1 for Useless comparison test
* 1 for Dereferenced variable may be null

**fixed alerts:**

* 7 for Useless comparison test
* 1 for Useless null check

---

*Comment posted by [LGTM.com](https://lgtm.com)*

cdjackson commented:
@hsudbrock I've refactored out the ```ZclCommandType``` enum. I've done some basic testing, but there are some issues with this approach which we can discuss next time we talk...

I'll look at client side attributes over the coming days.

hsudbrock commented:
@cdjackson 

Hi Chris; it looks to me like the autocoder does not compile with the current state of this PR (the file `ZigBeeZdoClusterGenerator.java` is missing). In case I don't overlook anything, could you add the missing file?

cdjackson commented:
Sorry Henning - I've now removed it. I would suggest for now not to rebase on the latest code anyway as there are still some issues I'm working through - hopefully I'll have this sorted in the next day or so.

hsudbrock commented:
>  I would suggest for now not to rebase on the latest code anyway as there are still some issues I'm working through - hopefully I'll have this sorted in the next day or so.

OK, thanks.

cdjackson commented:
This pull request **introduces 2 alerts** and **fixes 8** when merging 565b71ff4b1b3d8aaed42321c9444bd9163c03d2 into 0ebb115d2d595870b692a004ae154429ba6b7931 - [view on LGTM.com](https://lgtm.com/projects/g/zsmartsystems/com.zsmartsystems.zigbee/rev/pr-097912a7ab9a47d4e04b08abbcd99b9b90a0b028)

**new alerts:**

* 1 for Useless comparison test
* 1 for Dereferenced variable may be null

**fixed alerts:**

* 7 for Useless comparison test
* 1 for Useless null check

---

*Comment posted by [LGTM.com](https://lgtm.com)*

cdjackson commented:
This pull request **introduces 2 alerts** and **fixes 8** when merging cc0204a45f8df4cc3a8e7661877056b3bacf1ee7 into 0ebb115d2d595870b692a004ae154429ba6b7931 - [view on LGTM.com](https://lgtm.com/projects/g/zsmartsystems/com.zsmartsystems.zigbee/rev/pr-d0fc714c66458882ffce937a9883ab0a2692972a)

**new alerts:**

* 1 for Useless comparison test
* 1 for Dereferenced variable may be null

**fixed alerts:**

* 7 for Useless comparison test
* 1 for Useless null check

---

*Comment posted by [LGTM.com](https://lgtm.com)*

cdjackson commented:
This pull request **introduces 2 alerts** and **fixes 8** when merging a47ff1ba6d02785128d203380fac79c71e7b5921 into 0586e932c26d3bcac836fb1a5fd62cd0ddb5f139 - [view on LGTM.com](https://lgtm.com/projects/g/zsmartsystems/com.zsmartsystems.zigbee/rev/pr-a261a4761d007a7ea979b9a662d9fa41d653fa45)

**new alerts:**

* 1 for Useless comparison test
* 1 for Dereferenced variable may be null

**fixed alerts:**

* 7 for Useless comparison test
* 1 for Useless null check

---

*Comment posted by [LGTM.com](https://lgtm.com)*

cdjackson commented:
This pull request **introduces 2 alerts** and **fixes 8** when merging 25887e17309f78b816eb3428fe2ca311bb872f27 into 0586e932c26d3bcac836fb1a5fd62cd0ddb5f139 - [view on LGTM.com](https://lgtm.com/projects/g/zsmartsystems/com.zsmartsystems.zigbee/rev/pr-b5c235759bd5fa1d0d082fcdb7af282f91484702)

**new alerts:**

* 1 for Useless comparison test
* 1 for Dereferenced variable may be null

**fixed alerts:**

* 7 for Useless comparison test
* 1 for Useless null check

---

*Comment posted by [LGTM.com](https://lgtm.com)*

hsudbrock commented:
Hi Chris, one question: With this PR, you define the attributes `PIROccupiedToUnoccupiedDelay` and `PIRUnoccupiedToOccupiedDelay` from the occupancy sensing cluster `0x0406` as `uint8` (see [here](https://github.com/zsmartsystems/com.zsmartsystems.zigbee/blob/25887e17309f78b816eb3428fe2ca311bb872f27/com.zsmartsystems.zigbee.autocode/src/main/resources/0406_OccupancySensing.xml#L14) and [here](https://github.com/zsmartsystems/com.zsmartsystems.zigbee/blob/25887e17309f78b816eb3428fe2ca311bb872f27/com.zsmartsystems.zigbee.autocode/src/main/resources/0406_OccupancySensing.xml#L18)), whereas the ZSS spec I found specifies `uint16` for those two attributes.

Is there a reason for using 8 bits only, or should this be changed?

cdjackson commented:
I would need to check the spec, but if you say it should be a 16 but value then it’s quite possible I have made a mistake here :)

Sent from my iPhone

> On 3 Apr 2019, at 16:35, Henning Sudbrock <notifications@github.com> wrote:
> 
> Hi Chris, one question: With this PR, you define the attributes PIROccupiedToUnoccupiedDelay and PIRUnoccupiedToOccupiedDelay from the occupancy sensing cluster 0x0406 as uint8 (see here and here), whereas the ZSS spec I found specifies uint16 for those two attributes.
> 
> Is there a reason for using 8 bits only, or should this be changed?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

cdjackson commented:
This pull request **introduces 3 alerts** and **fixes 8** when merging ea3327eb63b1559aa7821f049d0e3c76bbe3a0ec into cfca6e7553cbc3b909ea312f47fc27ecaa846ebd - [view on LGTM.com](https://lgtm.com/projects/g/zsmartsystems/com.zsmartsystems.zigbee/rev/pr-55a9be4eb2f195576f9d2806b5e182020840dda1)

**new alerts:**

* 1 for Useless comparison test
* 1 for Dereferenced variable may be null
* 1 for Unused format argument

**fixed alerts:**

* 7 for Useless comparison test
* 1 for Useless null check

---

*Comment posted by [LGTM.com](https://lgtm.com)*

cdjackson commented:
This pull request **fixes 8 alerts** when merging 22f8c83f0de03a8467ebd71cda4e18bc45c5ffdc into cfca6e7553cbc3b909ea312f47fc27ecaa846ebd - [view on LGTM.com](https://lgtm.com/projects/g/zsmartsystems/com.zsmartsystems.zigbee/rev/pr-caa0fff2b3bef3b3134aee60aea4712178038f44)

**fixed alerts:**

* 7 for Useless comparison test
* 1 for Useless null check

---

*Comment posted by [LGTM.com](https://lgtm.com)*

cdjackson commented:
This pull request **fixes 8 alerts** when merging d47fce4d9f855760ea8c0e1fb41a5a0778da2bfd into 1671d475f56f9b2ddc839ab36e35a324bda2d854 - [view on LGTM.com](https://lgtm.com/projects/g/zsmartsystems/com.zsmartsystems.zigbee/rev/pr-a1b4cd8179d50c087d393e851cfa9c35dfc87316)

**fixed alerts:**

* 7 for Useless comparison test
* 1 for Useless null check

---

*Comment posted by [LGTM.com](https://lgtm.com)*

cdjackson commented:
This pull request **fixes 8 alerts** when merging 5c4c58716329b4f66e48595f85be9b33fbc4d904 into edcce461b44da56de86e7d0f80a5e2b593b62013 - [view on LGTM.com](https://lgtm.com/projects/g/zsmartsystems/com.zsmartsystems.zigbee/rev/pr-d37efc4319d6a383146c6e75cf592efea0ee480c)

**fixed alerts:**

* 7 for Useless comparison test
* 1 for Useless null check

---

*Comment posted by [LGTM.com](https://lgtm.com)*

cdjackson commented:
This pull request **fixes 8 alerts** when merging d364c8ee030539b7757b765fd24326131aa03d92 into f14fc5083965d87eef2786671781e257fdd2e0bc - [view on LGTM.com](https://lgtm.com/projects/g/zsmartsystems/com.zsmartsystems.zigbee/rev/pr-e1f4a18996c1fa63dfd432d02405d3a5ad12d101)

**fixed alerts:**

* 7 for Useless comparison test
* 1 for Useless null check

---

*Comment posted by [LGTM.com](https://lgtm.com)*
